src/arch/x86/reg/amd64_core.rs

```diff
@@ -0,0 +1,155 @@
+use crate::arch::Registers;
+use core::convert::TryInto;
+
+/// 64-bit x86 core registers.
+///
+/// Source: https://github.com/bminor/binutils-gdb/blob/master/gdb/features/arm/arm-core.xml
+#[derive(Default)]
+pub struct Amd64CoreRegs {
+    // RAX, RBX, RCX, RDX, RSI, RDI, RBP, RSP, r8-r15
+    regs: [u64; 16],
+    // Status register
+    eflags: u32,
+    // Instruction pointer
+    rip: u64,
+    // Segment registers: CS, SS, DS, ES, FS, GS
+    segments: [u32; 6],
+    // FPU registers: ST0 through ST7
+    st_regs: [F80; 8],
+    // SIMD Registers: XMM0 through XMM15
+    xmm_regs: [u128; 0x10],
+    // SSE Status/Control Register
+    mxcsr: u32,
+}
+
+type F80 = [u8; 10];
+
+impl Registers for Amd64CoreRegs {
+    fn gdb_serialize(&self, mut write_byte: impl FnMut(Option<u8>)) {
+        macro_rules! write_bytes {
+            ($bytes:expr) => {
+                for b in $bytes {
+                    write_byte(Some(*b))
+                }
+            };
+        }
+
+        for reg in &self.regs {
+            write_bytes!(&reg.to_le_bytes());
+        }
+
+        // rip
+        write_bytes!(&self.rip.to_le_bytes());
+
+        // eflags
+        write_bytes!(&self.eflags.to_le_bytes());
+
+        // cs, ss, ds, es, fs, gs
+        for seg in &self.segments {
+            write_bytes!(&seg.to_le_bytes());
+        }
+
+        // st0 to st7
+        for st_reg in &self.st_regs {
+            write_bytes!(st_reg);
+        }
+
+        // fctrl, fstat, ftag, fiseg, fioff, foseg, fooff, fop
+        for _ in 0..8 {
+            write_bytes!(&[0; 4]);
+        }
+
+        // xmm0 to xmm15
+        for xmm_reg in &self.xmm_regs {
+            write_bytes!(&xmm_reg.to_le_bytes());
+        }
+
+        // mxcsr
+        write_bytes!(&self.mxcsr.to_le_bytes());
+
+        // padding?
+        write_bytes!(&[0x10; 0x18]);
+    }
+
+    fn gdb_deserialize(&mut self, bytes: &[u8]) -> Result<(), ()> {
+        let mut regs = bytes[0..0x80]
+            .chunks_exact(8)
+            .map(|x| u64::from_le_bytes(x.try_into().unwrap()));
```

Comment:
In this situation, the unwrap panic is unreachable without a serious error on the part of the programmer, as both the indexing and `chunks_exact` would cause issues before that unwrap is reachable in situations which would cause a panic.

I have, however, done my best to remove all the unwraps, however I wasn't able to get the last one (added alongside all the nasty FPU internal registers) without making the code significantly more verbose.

I honestly assumed this style of unwrapping was fine due to it being used [elsewhere in the codebase](https://github.com/daniel5151/gdbstub/blob/master/src/arch/arm/reg/arm_core.rs#L54)

Reply:
Ah, you know what, I'm a dummy. 
You're totally right, and in-fact, this is the one unwrap case that totally makes sense!

What I meant to convey was unwrapping in the _rest_ of the function wasn't necessary. That's on me 😅

Let's re-add that key unwrap, and make the rest of the code a lot less verbose, yeah? That aught to let us remove all those nasty `map_err(|_|())?` littered about.